src/Renderer/Rectangle.cpp

```diff
@@ -319,26 +279,24 @@ bool Rectangle_2df::operator==(const Rectangle_2df& rect)
 /**
  * Compares two rectangles.
  *
- * \param rect	Rectangle_2d to compare against.
+ * \param rect	Rectangle_2df to compare against.
  *
  * \returns True if the two rectangles are not equivalent.
  */
-bool Rectangle_2df::operator!=(const Rectangle_2d& rect)
+bool Rectangle_2df::operator!=(const Rectangle_2df& rect) const
 {
 	return (x() != rect.x()) || (y() != rect.y()) || (width() != rect.width()) || (height() != rect.height());
 }
 
 
-/**
- * Compares two rectangles.
- *
- * \param rect	Rectangle_2df to compare against.
- *
- * \returns True if the two rectangles are not equivalent.
- */
-bool Rectangle_2df::operator!=(const Rectangle_2df& rect)
+Rectangle_2df::operator Rectangle_2d() const
 {
-	return (x() != rect.x()) || (y() != rect.y()) || (width() != rect.width()) || (height() != rect.height());
+	return Rectangle_2d{
+		static_cast<int>(mX),
+		static_cast<int>(mY),
+		static_cast<int>(mW),
+		static_cast<int>(mH)
+	};
```

Comment:
```
#include <cmath>
//...
return Rectangle_2d {
static_cast<int>(std::floor(mX))
static_cast<int>(std::floor(mY))
static_cast<int>(std::floor(mW))
static_cast<int>(std::floor(mH))
};
```
Floor float values before casting to int to prevent unintended behavior when dealing with negative values.
(You'll need to include `<cmath>`)

Note the graph and examples here; if you only truncate you get -2.0 instead of -3.0 when dealing with negative values.
Ref: https://en.cppreference.com/w/cpp/numeric/math/floor

Reply:
You know, I was tempted to leave that method out entirely, since the original didn't have auto conversion in that direction. I eventually decided to include it for symmetry, and because it seemed to fit with the idea of templatizing things later on.

----

I think delving into rounding modes is starting to get a bit subjective. I would not consider the proposed behaviour to be very intuitive as a default. It does seem potentially useful, and I could see us supporting that. I think though, due to the various forms of rounding, we would need to start using named methods to choose the proper rounding mode. The programmer could then explicitly choose which rounding mode they want.

```cpp
Rect<int> rectIntFloor = rectFloat.floor();
Rect<int> rectIntCeil = rectFloat.ceil();
Rect<int> rectIntRound = rectFloat.round();
Rect<int> rectIntTrunc = rectFloat.trunc();
```

Note that the standard library floating point rounding methods return floating point results. Hence the rounding method on the `Rect` may actually return a `Rect<float>` still, rather than a `Rect<int>`. In the above code, the assignment might then do an implicit cast from `Rect<float>` (of the temporary result) to `Rect<int>`.

For the default conversion operator, which can be triggered with `static_cast` of the struct, I think the most intuitive way to implement it, is to apply the `static_cast` to each field using the underlying types of the two structs. In particular, the reverse operation from `int` to `float` is implemented that way.

```cpp
// Applies `static_cast<float>` to each member field
Rect<float> rectFloat = static_cast<Rect<float>>(rectInt);
```

If both directions were consistent, then when we templatize the class, both directions can collapse into a single template method. That's one reason why I feel using a simple `static_cast` of each member field is most intuitive here.

```cpp
Rect<int> rectIntExplicitCast = static_cast<Rect<int>>(rectFloat); // Equivalent to truncation
Rect<int> rectIntImplicitConversion = rectFloat; // Equivalent to truncation
```

----

My suggestion here, is to remain with a simple `static_cast` for this operation. I would however quite welcome another issue to offer support for other rounding modes.

Reply:
My original thought of "do as math does" should be the default "intuitive" implementation but then I found out apparently "doing what math does" is just as ambiguous!

Retracting my requested changes.

Ref: https://math.stackexchange.com/questions/3448/rules-for-rounding-positive-and-negative-numbers
Ref: https://en.wikipedia.org/wiki/Rounding

Reply:
Huh, that Wikipedia article looks like a pretty good summary of rounding.